### PR TITLE
[FLINK-23767][streaming] FLIP-180: Adjust StreamStatus and Idleness definition.

### PR DESCRIPTION
**Pull request by AHeise**
## What is the purpose of the change

StreamStatus is currently defined as follows:

>    A Stream Status element informs stream tasks whether or not they should continue to expect records and watermarks from the input stream that sent them. There are 2 kinds of status, namely IDLE and ACTIVE. Stream Status elements are generated at the sources, and may be propagated through the tasks of the topology. They directly infer the current status of the emitting task; a SourceStreamTask or StreamTask emits a IDLE if it will temporarily halt to emit any records or watermarks (i.e. is idle), and emits a ACTIVE once it resumes to do so (i.e. is active).

However that definition is quite problematic. Operators with delayed responses, such as AsyncIO, windows, or time-based process functions, might output elements when they are actually idle. So by that definition, they would need to switch from idle to active, emit an element, and go back to idle. It's not entirely clear when they would go back to idle (after each element or after some time?). Furthermore, under heavy backpressure, these additional StreamStatuses delay progress quite a bit.

At this point, it makes sense to revise the original intent of StreamStatus: being able to advance the watermark without data. Indeed, there is no reason to expand the definition towards records; we could simply limit the definition to watermarks as this is the only use case for StreamStatus in the code base currently. Moreover, that's exactly how it's implemented in Flink so far. So, in this FLIP we propose to adjust the definition to reality and act on that.

This is the implementation PR for [FLIP-180](https://cwiki.apache.org/confluence/display/FLINK/FLIP-180%3A+Adjust+StreamStatus+and+Idleness+definition).

## Brief change log

- Update definition of StreamStatus.
- Rename StreamStatus to WatermarkStatus.
- Add WatermarkOutput#markActive. 

## Verifying this change

Mostly renaming that is covered by existing tests.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (**yes** / no)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (**yes** / no / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (not applicable / docs / **JavaDocs** / not documented)

### CODE REVIEW

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/watermarkstatus/WatermarkStatus.java [hunk 1]
```diff
@@ -28,11 +28,11 @@
 /**
  * A Stream Status element informs stream tasks whether or not they should continue to expect
```
Comment (fapaul) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/watermarkstatus/WatermarkStatus.java:28]: Unfortunately, it seems that we also use `Stream status` in our codebase you probably also need to replace those occurrences.

File: flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/internals/SourceContextWatermarkOutputAdapter.java [hunk 2]
```diff
@@ -42,4 +42,9 @@ public void emitWatermark(Watermark watermark) {
     public void markIdle() {
         sourceContext.markAsTemporarilyIdle();
     }
+
+    @Override
+    public void markActive() {
+        // will be set active with next watermark
```
Comment (fapaul) [inline flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/internals/SourceContextWatermarkOutputAdapter.java:48]: Do we need some kind of follow-up to implement this?

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/source/WatermarkToDataOutput.java [hunk 3]
```diff
@@ -83,4 +80,25 @@ public void markIdle() {
...
+    @Override
+    public void markActive() {
+        try {
+            markActiveInternally();
+        } catch (ExceptionInChainedOperatorException e) {
+            throw e;
+        } catch (Exception e) {
+            throw new ExceptionInChainedOperatorException(e);
+        }
+    }
```
Comment (fapaul) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/source/WatermarkToDataOutput.java:93]: Why do we handle the exceptions differently than before?

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/watermarkstatus/StatusWatermarkValve.java [hunk 4]
```diff
@@ -1,11 +1,10 @@
 /*
- * Licensed to the Apache Software Foundation (ASF) under one
- * or more contributor license agreements.  See the NOTICE file
- * distributed with this work for additional information
- * regarding copyright ownership.  The ASF licenses this file
- * to you under the Apache License, Version 2.0 (the
- * "License"); you may not use this file except in compliance
- * with the License.  You may obtain a copy of the License at
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
```
Comment (fapaul) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/watermarkstatus/StatusWatermarkValve.java:3]: The formatting changes do not belong to this commit.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/watermarkstatus/WatermarkStatus.java [hunk 5]
```diff
@@ -1,11 +1,10 @@
 /*
- * Licensed to the Apache Software Foundation (ASF) under one
- * or more contributor license agreements.  See the NOTICE file
- * distributed with this work for additional information
- * regarding copyright ownership.  The ASF licenses this file
- * to you under the Apache License, Version 2.0 (the
- * "License"); you may not use this file except in compliance
- * with the License.  You may obtain a copy of the License at
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
```
Comment (fapaul) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/watermarkstatus/WatermarkStatus.java:3]: Same here wrong commit.

**Review by fapaul — COMMENTED** [review]
Overall it looks already quite good.

I think you have missed some occurrences during the renaming but I also used my IDE and did not find `[S,s]tream[S,s]status` in the codebase anymore 👍

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/watermarkstatus/WatermarkStatus.java (hunk 1 above)
Comment (AHeise) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/watermarkstatus/WatermarkStatus.java:28]: I added 3 more grep patterns. Good catch!

File: flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/internals/SourceContextWatermarkOutputAdapter.java (hunk 2 above)
Comment (AHeise) [reply flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/internals/SourceContextWatermarkOutputAdapter.java:48]: This is the old interface; so I'd like to avoid it.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/source/WatermarkToDataOutput.java (hunk 3 above)
Comment (AHeise) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/source/WatermarkToDataOutput.java:93]: This method did not exist before; I copied the handling from #markIdle and #emitWatermark without thinking too much tbh.

File: flink-core/src/main/java/org/apache/flink/api/common/eventtime/CombinedWatermarkStatus.java [hunk 6]
```diff
@@ -69,7 +69,6 @@ public boolean updateCombinedWatermark() {
         // if we don't have any outputs minimumOverAllOutputs is not valid, it's still
         // at its initial Long.MAX_VALUE state and we must not emit that
         if (partialWatermarks.isEmpty()) {
-            this.idle = combinedWatermark > Long.MIN_VALUE;
```
Comment (fapaul) [inline flink-core/src/main/java/org/apache/flink/api/common/eventtime/CombinedWatermarkStatus.java:72]: Do we want to provide some default Watermark strategy to mimic this behavior? This can be probably done through a follow-up.

**Review by fapaul — APPROVED** [review]
LGTM 👍  I could not find any stream status occurrences in the codebase besides Kinesis' definitions. I wonder if someone from the more runtimy teams should also take a quick look.

File: flink-core/src/main/java/org/apache/flink/api/common/eventtime/CombinedWatermarkStatus.java (hunk 6 above)
Comment (AHeise) [reply flink-core/src/main/java/org/apache/flink/api/common/eventtime/CombinedWatermarkStatus.java:72]: Good question; I'm also not sure if it's trivial to implement on a higher level... I'm almost assuming that source implementors need to implement it themselves. But afaik we also don't encourage it. @dawidwys WDYT?
